File: remove-stones-to-minimize-the-total/remove-stones-to-minimize-the-total.java
class Solution {
    public int minStoneSum(int[] piles, int k) {
        
        PriorityQueue<Integer> pQ = new PriorityQueue<>((a, b) -> (b - a));
        
        int sum = 0;
        
        for (int p : piles) {
            
            sum += p;
            
            pQ.add(p);
        }
        
        while (k > 0) {
            
            int temp = pQ.poll();
            
            sum -= (temp / 2);
            
            pQ.add(temp - (temp / 2));
            
            --k;
        }
        
        return sum;
    }
}

File: remove-stones-to-minimize-the-total/remove-stones-to-minimize-the-total.cpp
class Solution {
public:
    int minStoneSum(vector<int>& piles, int k) {
        
        priority_queue<int> pQ(piles.begin(), piles.end());
        
        for (int i = 0; i < k; ++i) {
            
            int temp = pQ.top();
            pQ.pop();
            pQ.push(temp - (temp / 2));
        }
        
        int min = 0;
        
        while (!pQ.empty()) {
            
            min += pQ.top();
            pQ.pop();
        }
        
        return min;
    }
};